ocaml/xapi/repository.ml

```diff
@@ -720,40 +668,41 @@ let apply_updates' ~__context ~host ~updates_info ~livepatches ~acc_rpm_updates
             )
         )
       ]
-      ) ;
-  (* Evaluate guidances *)
-  let guidances =
-    let guidances' =
-      (* EvacuateHost will be applied before applying updates *)
-      eval_guidances ~updates_info ~updates:acc_rpm_updates ~kind:Recommended
-        ~livepatches:successful_livepatches ~failed_livepatches
-      |> List.filter (fun g -> g <> Guidance.EvacuateHost)
-      |> fun l -> merge_with_unapplied_guidances ~__context ~host ~guidances:l
-    in
-    GuidanceSet.assert_valid_guidances guidances' ;
-    match failed_livepatches with
-    | [] ->
-        guidances'
-    | _ :: _ ->
-        (* There is(are) livepatch failure(s):
-         * the host should not be rebooted, and
-         * an extra pending guidance 'RebootHostOnLivePatchFailure' should be set.
-         *)
-        guidances'
-        |> List.filter (fun g -> g <> Guidance.RebootHost)
-        |> List.cons Guidance.RebootHostOnLivePatchFailure
+      )
+
+let apply_updates' ~__context ~host ~updates_info ~livepatches ~acc_rpm_updates
+    =
+  let open Guidance in
+  (* This function runs on coordinator host *)
+  let get_guidances kind =
+    eval_guidances ~updates_info ~updates:acc_rpm_updates ~kind ~livepatches
+    |> GuidanceSet.remove EvacuateHost
+    |> GuidanceSet.elements
   in
+  let mandatory = get_guidances Mandatory in
+  (* Install RPM updates firstly *)
+  Helpers.call_api_functions ~__context (fun rpc session_id ->
+      Client.Client.Repository.apply ~rpc ~session_id ~host
+  ) ;
+  (* Apply livepatches *)
+  let _, failed_livepatches =
+    match List.mem RebootHost mandatory with
+    | true ->
+        (* Not apply any livepatches as the host will reboot *)
+        ([], [])
+    | false ->
+        apply_livepatches' ~__context ~host ~livepatches
+  in
+  (* Update states in cache *)
+  update_cache ~host ~failed_livepatches ;
   List.iter
-    (fun g -> debug "pending_guidance: %s" (Guidance.to_string g))
-    guidances ;
-  set_pending_guidances ~__context ~host ~guidances ;
-  ( guidances
-  , List.map
-      (fun (lp, _) ->
-        [Api_errors.apply_livepatch_failed; LivePatch.to_string lp]
-      )
-      failed_livepatches
-  )
+    (fun g -> debug "mandatory pending_guidance: %s" (to_string g))
+    mandatory ;
+  let ops = get_ops_of_pending ~__context ~host ~kind:Mandatory in
+  set_pending_guidances ~ops ~coming:mandatory ;
```

Comment:
Only mandatory guidances are set? When will recommended and full guidances be set?

Reply:
As described in commit message, others will be set in next PR.